server/src/controller/oauth2.ts

```diff
@@ -36,17 +39,48 @@ export default class OAuth2Controller {
      */
     //todo: jwt verification intermediate step
     //todo: make POST operation
-    public static async tokenExchange(ctx: BaseContext): Promise<void> {
+    public static async tokenExchange(ctx: Context): Promise<void> {
         const oauthCode = ctx.query["code"];
-        const {tokens} = await oauth2Client.getToken(oauthCode)
-        const id_token: any = jwt.decode(tokens.id_token); //todo: make interface
+        const {tokens} = await oauth2Client.getToken(oauthCode);
         OAuth2Controller.setCookies(ctx, tokens);
-        ctx.body = id_token.given_name;
+        const ticket = await OAuth2Controller.verifyToken(tokens.id_token);
+        const user: User = await OAuth2Controller.getOrCreateUser(ticket.payload)
+        ctx.body = user.name;
+    }
+
+    public static async verifyToken(idToken: string): Promise<any> {
+        try {
+            return await oauth2Client.verifyIdToken({
+                idToken: idToken,
+                audience: keys.YOUR_CLIENT_ID
+            });
+        } catch {
+            return null;
+        }
+    }
```

Comment:
It's for code reuse, both the middleware and the controller use it. It should technically be part of the service layer. If I don't have this method then I would have to write the same unit test in the middleware controller rather than just stubbing it.